### ISSUE TITLE
magit incorrectly treats file as untracked if file has executable bit set

### ISSUE THREAD

**Issue by zuav** [magit/magit#2408]
Magit 20151123.1028, Git 2.6.3, Emacs 24.5.1 OS X 10.11.1

It's very easy to reproduce the bug.
Create a new repo with one bash script in it, set executable bit on the script and commit it to the repo. Then try magic-log-buffer-file on the script file.

**Comment by kyleam** [comment]
Hmm, I'm not able to reproduce this on a GNU/Linux system (Emacs 25).

**Comment by zuav** [comment]
If you unable to reproduce the bug on Linux than I guess it's BSD specific.
I've just checked that if executable bit is cleaned than everything is OK.
When I eval in the script buffer than I get the following:

(magit-file-relative-name buffer-file-name) => nil
(magit-file-tracked-p (file-relative-name buffer-file-name)) => t

I've looked at the magic-file-relative-name code and understood little ;)
Do you have access to any BSD machine?

**Comment by kyleam** [comment]
No, sorry.  I'm hoping others, especially OS X users, will report
whether they can reproduce it.

Perhaps you can step through magit-file-relative-name with the +x and
-x versions of the file, and see where the execution diverges.

**Comment by johnmastro** [comment]
I can't reproduce it on OS X 10.11 with either the default `git` (2.4.9) or 2.6.3 from homebrew.

@zuav, are there symlinks involved? This issue sounds familiar, but I haven't found the other issue I'm thinking of.

Also, is it safe to assume that if you evaluate this form via `M-:`, from the file in question, it returns `nil`?

```emacs-lisp
(magit-toplevel (magit--with-safe-default-directory buffer-file-name
                  default-directory))
```

**Comment by zuav** [comment]
@johnmastro, it's strange, I thought my setup is pretty straightforward.
No, there no symlinks involved.

(magit-file-relative-name buffer-file-name) => nil
(magi-file-relative-name) => nil
(buffer-file-name) => "/private/var/tmp/repo/script.sh"
(file-relative-name buffer-file-name) => "script.sh"
(magit-file-tracked-p (file-relative-name buffer-file-name)) => t
(magit-toplevel (magit--with-safe-default-directory buffer-file-name default-directory)) => nil

**Comment by johnmastro** [comment]
Will you try this too, again via `M-:' from the file?

```emacs-lisp
(magit--with-safe-default-directory buffer-file-name default-directory)
```

**Comment by johnmastro** [comment]
A few other things to try:

```emacs-lisp
(magit-rev-parse-safe "--show-toplevel")
(magit-rev-parse-safe "--git-dir")
```

And from the shell:

```sh
ls -l /private/var/tmp/repo
ls -l /private/var/tmp/repo/script.sh
```

**Comment by zuav** [comment]
Here:

(magit--with-safe-default-directory buffer-file-name default-directory) => "/private/var/tmp/repo/script.sh/"
(magit-rev-parse-safe "--show-toplevel") => "/private/var/tmp/repo"
(magit-rev-parse-safe "--git-dir") => ".git"

zuav@ithilien$ LC_ALL=C ls -al /private/var/tmp/repo
total 8
drwxr-xr-x   4 zuav  wheel  136 Nov 27 00:11 .
drwxrwxrwt   9 root  wheel  306 Nov 26 20:37 ..
drwxr-xr-x  13 zuav  wheel  442 Nov 25 21:50 .git
-rwxr-xr-x   1 zuav  wheel  583 Nov 27 00:11 script.sh

**Comment by tarsius** [comment]
For increased readability please wrap command output, see https://guides.github.com/features/mastering-markdown/#examples.

**Comment by tarsius** [comment]
So it seems the issue is in `magit-toplevel`. If you know elisp, then please sprinkle some `message`s in there to see where things start going down the drain.

**Comment by johnmastro** [comment]
@tarsius, I haven't had a chance to look at this more but I noticed that he shows the `default-directory` is reported as `""/private/var/tmp/repo/script.sh/"` - the name of the file with a slash after it. If I understand correctly, no such directory exists on his system.

I don't know if that's significant, and I couldn't reproduce any issues by mimicking it, but it does seem strange.

**Comment by npostavs** [comment]
> I haven't had a chance to look at this more but I noticed that he shows the default-directory is reported as ""/private/var/tmp/repo/script.sh/" - the name of the file with a slash after it.

Hmm, perhaps `file-accessible-directory-p` returns `t` for executable files when they're named as directories on OSX (doesn't happen on GNU/Linux)?

**Comment by zuav** [comment]
Yes, ```file-accessible-directory-p``` reports incorrect results here.
```
(file-accessible-directory-p "/private/var/tmp/repo/script.sh") => t

**Comment by tarsius** [comment]
Also see https://github.com/magit/magit/issues/2295.

**Comment by npostavs** [comment]
> Also see #2295.

Ah, this does appear to be the same root cause. No activity on the Emacs bug though http://debbugs.gnu.org/cgi/bugreport.cgi?bug=21573

Can we work around it by doing `(and (file-directory-p file) (file-accessible-directory-p file))`, or is `file-directory-p` broken as well?

**Comment by kyleam** [comment]
Doh.  Sorry everyone that I didn't recall that issue.  Should've have rang a bell right away.

**Comment by johnmastro** [comment]
> Also see #2295.

Yikes, so it's even nondeterministic?

Like kyleam, I tried invoking `file-accessible-directory-p` repeatedly (using the loop below), but never got any result but `nil`. My results were the same on both 25.0.50 and 24.5.1.

```emacs-lisp
(defvar file-accessible-directory-p--results nil)

(let* ((file-name "/private/var/tmp/repo/script.sh")
       (dir-name (file-name-as-directory file-name)))
  (setq file-accessible-directory-p--results nil)
  (dotimes (_ 100)
    (add-to-list 'file-accessible-directory-p--results
                 (list (file-accessible-directory-p file-name)
                       (file-accessible-directory-p dir-name)))
    (sit-for (/ (random 5) 10.0)))
  (message "-- file-accessible-directory-p--results: %s"
           file-accessible-directory-p--results))
```

@zuav, to try npostavs's suggestion, evaluate this and see what result you get

```emacs-lisp
(file-directory-p "/private/var/tmp/repo/script.sh")
```

**Comment by zuav** [comment]
@johnmastro, I got with your code ```"-- file-accessible-directory-p--results: ((t t))"```
and ```file-directory-p``` evaluated to nil:
```
(file-directory-p "/private/var/tmp/repo/script.sh") => nil
```
I do not think this is Emacs bug though.  When I start Emacs like this: ```emacs -Q``` then
```
(file-accessible-directory-p "/private/var/tmp/repo/script.sh") => nil
```
and
```
"-- file-accessible-directory-p--results: ((nil nil))"
```

**Comment by johnmastro** [comment]
> file-directory-p evaluated to nil

Then the good news is that @npostavs's proposed workaround will probably work, regardless of what's actually causing the bug.

> I do not think this is Emacs bug though

On issue #2295, the problem reproduced even in `emacs -Q`, so I think it remains most likely that it is an Emacs bug. However, I can think of two ways a package could affect `file-accessible-directory-p`: advice and file handlers.

To check for those:

1. Use `describe-function` on `file-accessible-directory-p` (e.g. `C-h f file-accessible-directory-p RET`), and see if it mentions anything about advice.
2. Evaluate the code below to see if any handlers match the file name in question.

```emacs-lisp
(cl-remove-if-not
 (lambda (pair)
   (string-match-p (car pair) "/private/var/tmp/repo/script.sh"))
 file-name-handler-alist)
```

**Comment by zuav** [comment]
Here:
```describe-function``` mentions nothing about advice

```lisp 
(cl-remove-if-not
  (lambda (pair)
    (string-match-p (car pair) "/private/var/tmp/repo/script.sh"))
  file-name-handler-alist)  => nil
```
```lisp
(find-file-name-handler "/private/var/tmp/repo/script.sh"
                                     'file-accessible-directory-p)  => nil
```

**Comment by zuav** [comment]
One more point.
I did run ```emacs -Q``` from terminal window.
If I run ```emacs``` from terminal window without ```-Q``` option it still evaluates correctly
```
(file-accessible-directory-p "/private/var/tmp/repo/script.sh") => nil
```
It seems that something wrong with Emacs.app.

**Comment by npostavs** [comment]
> I did run emacs -Q from terminal window.

Can you clarify if the problem occurs with `emacs -Q` in graphical mode (and add this info to [Emacs bug #21573][21573])?

[21573]: http://debbugs.gnu.org/cgi/bugreport.cgi?bug=21573.

**Comment by zuav** [comment]
Funny. Here's new steps to reproduce the bug. I hope somebody will be able to repeat the bug.
*  Open fresh terminal, I mean close running terminal app with Cmd-Q and run a fresh copy
*  Run Emacs app like this: ```/Applications/Emacs.app/Contents/MacOS/Emacs -Q &```
* Eval in the scratch buffer ```(file-accessible-directory-p "/private/var/tmp/repo/script.sh")```. The result should  be```nil``` 
* Open file: C-x C-f "/private/var/tmp/repo/script.sh"
* Go back the scratch buffer and eval ```(file-accessible-directory-p "/private/var/tmp/repo/script.sh")```. The result should be ```t```
* Exit Emacs with C-x C-c
* Run Emacs from the same terminal window
* Eval in the scratch buffer ```(file-accessible-directory-p "/private/var/tmp/repo/script.sh")```. The result should be ```t``` 

 NB I'm using standard terminal app that goes with OS X.

**Comment by npostavs** [comment]
> Funny. Here's new steps to reproduce the bug. I hope somebody will be able to repeat the bug.

Please send that to 21573@debbugs.gnu.org.

Meanwhile, does 6f2980b0187be94f45b5dd8b4245a176fe5991b7 avoid the problem?

**Comment by zuav** [comment]
@npostavs, https://github.com/magit/magit/commit/6f2980b0187be94f45b5dd8b4245a176fe5991b7 helps, thank you.

I did sent the description. But I wonder will somebody lucky enough to reproduce the situation?

**Comment by npostavs** [comment]
> But I wonder will somebody lucky enough to reproduce the situation?

Well, 2 people have already bumped into it, the question is will someone who is willing to debug it be able to reproduce it...

**Comment by npostavs** [comment]
Should be fixed in master now (05572d7110fcbd13a378bcebd40eb55b75b45a96).
